### PR TITLE
Allowing labeling of volume claim templates.

### LINKED ISSUE #7910: Adding a metadata label to volumeClaimTemplates does not work
## Bug Report
**What did you do?**

Edit the ES CR to add a label to the volumeclaimtemplate

**What did you expect to see?**

The label should be added to the exisitn pvcs

**What did you see instead? Under which circumstances?**

```
k edit elasticsearches.elasticsearch.k8s.elastic.co es-visibility 
error: elasticsearches.elasticsearch.k8s.elastic.co "es-visibility" could not be patched: admission webhook "elastic-es-validation-v1.k8s.elastic.co" denied the request: Elasticsearch.elasticsearch.k8s.elastic.co "es-visibility" is invalid: spec.nodeSet[0].volumeClaimTemplates: Invalid value: []v1.PersistentVolumeClaim{v1.PersistentVolumeClaim{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"elasticsearch-data", GenerateName:"", Namespace:"", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), DeletionTimestamp:<nil>, DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string{"velero.io/exclude-from-backup":"true"}, Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ManagedFields:[]v1.ManagedFieldsEntry(nil)}, Spec:v1.PersistentVolumeClaimSpec{AccessModes:[]v1.PersistentVolumeAccessMode{"ReadWriteOnce"}, Selector:(*v1.LabelSelector)(nil), Resources:v1.VolumeResourceRequirements{Limits:v1.ResourceList(nil), Requests:v1.ResourceList{"storage":resource.Quantity{i:resource.int64Amount{value:53687091200, scale:0}, d:resource.infDecAmount{Dec:(*inf.Dec)(nil)}, s:"50Gi", Format:"BinarySI"}}}, VolumeName:"", StorageClassName:(*string)(0xc00458b670), VolumeMode:(*v1.PersistentVolumeMode)(nil), DataSource:(*v1.TypedLocalObjectReference)(nil), DataSourceRef:(*v1.TypedObjectReference)(nil), VolumeAttributesClassName:(*string)(nil)}, Status:v1.PersistentVolumeClaimStatus{Phase:"", AccessModes:[]v1.PersistentVolumeAccessMode(nil), Capacity:v1.ResourceList(nil), Conditions:[]v1.PersistentVolumeClaimCondition(nil), AllocatedResources:v1.ResourceList(nil), AllocatedResourceStatuses:map[v1.ResourceName]v1.ClaimResourceStatus(nil), CurrentVolumeAttributesClassName:(*string)(nil), ModifyVolumeStatus:(*v1.ModifyVolumeStatus)(nil)}}}: volume claim templates can only have their storage requests increased, if the storage class allows volume expansion. Any other change is forbidden

```

**Environment**

* ECK version:

	2.13.0

* Kubernetes information:

	insert any information about your Kubernetes environment that could help us:
	* Cloud:  AKS 1.28.5 ?

	
	for each of them please give us the version you are using
```
Client Version: v1.30.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.28.5

```

### PR DESCRIPTION
**Pull request by naemono**
# Allow labeling of volume claim templates

Resolves https://github.com/elastic/cloud-on-k8s/issues/7910

## Why this change

Kubernetes StatefulSets treat `spec.volumeClaimTemplates` as largely immutable. Users still need labels on the backing PVCs (for example Velero exclusions, cost allocation, or team ownership). Today those labels could not be expressed on the CR’s volume claim templates and applied reliably without fighting apiserver errors or losing parity between desired state and actual PVCs.

This PR lets users set **non-reserved** labels on volume claim templates for **Elasticsearch** and **Logstash**, validates them at admission, and **propagates** them onto existing PVCs during reconciliation—without weakening existing guards around storage, claim names, or ECK-owned labels.

## Behavior contract (what reviewers should verify)

| Topic | Intended behavior |
|--------|-------------------|
| **StatefulSet update** | On an existing StatefulSet, reconciliation keeps the **already persisted** `volumeClaimTemplates` slice when patching the StatefulSet, so we do not trigger apiserver “field is immutable” failures for label-only VCT edits. Desired template labels still drive PVC updates via `ReconcilePVCsForStatefulSet`. |
| **PVC labels** | Sync from the **expected** claim template labels to each matching PVC is **additive for user keys**: new/changed keys from the CR are applied; clearing a label on the VCT does **not** remove that key from PVCs (documented in code/tests—matches “don’t surprise-delete user state”). |
| **Reserved keys** | Label keys owned by ECK (`metadata/reserved`, e.g. `*.k8s.elastic.co`, `eck.k8s.alpha.elastic.co/...`) are **rejected** on VCT create/update, **stripped** when building StatefulSet VCTs (no-webhook path), and **not** synced onto PVCs. The same rules apply to `propagate-labels` / `propagate-annotations: *` on parent CRs (see below). |
| **Bypass / no webhook** | `StripReservedLabelKeys` runs when building the StatefulSet (ES: `nodespec/statefulset.go`, Logstash: `reconcile.go`) so reserved keys cannot slip onto **new** PVCs via VCT copy-at-create if admission was skipped. |
| **Scale-up** | New PVCs may be created by the StatefulSet controller before the next reconcile pass applies labels from `ReconcilePVCsForStatefulSet`; labels converge on the following reconcile (comment on `sset.ReconcileStatefulSet`). |

## Metadata propagation (behavior change since 3.1.0)

This PR also centralizes ECK-owned metadata rules in `pkg/controller/common/metadata/reserved` and uses them for **volume claim template** validation, **PVC label sync**, and **[label/annotation propagation](https://github.com/elastic/cloud-on-k8s/pull/8673)** (#2652, shipped in 3.1.0).

That fixes an inconsistency called out in review: VCT reserved-key checks did not match propagation filtering (e.g. `eck.k8s.alpha.elastic.co/...`).

### What changes for propagation

When a parent CR has `eck.k8s.alpha.elastic.co/propagate-labels: "*"` or `propagate-annotations: "*"`:

| Metadata | Before (3.1.0, #8673) | After this PR |
|----------|------------------------|---------------|
| User labels/annotations (`team`, `velero.io/...`, etc.) | Propagated | Propagated (unchanged) |
| ECK label keys on the CR (`elasticsearch.k8s.elastic.co/...`, `common.k8s.elastic.co/...`, `eck.k8s.alpha.elastic.co/...`) | **Also propagated** with `propagate-labels: *` | **Excluded** |
| ECK annotation keys with `propagate-annotations: *` | Mostly excluded (regex) | Excluded via the same rules as labels; `k8s.elastic.co/...` annotations no longer slip through |

Explicit lists (`propagate-labels: "team,cost-center"`) are unchanged for allowed keys. Reserved keys are not copied even if named in the list (they are removed from the propagatable set).

### Why treat this as a breaking change

- **Scope:** Any cluster using `propagate-labels: *` since 3.1.0 and relying on ECK identity labels from the **parent CR** appearing on child objects via propagation only (not from operator reconciliation).
- **Severity:** Low for typical use (org/custom labels). Child resources still receive ECK labels from the operator where it already sets them.
- **Mitigation:** Use explicit `propagate-labels` / `propagate-annotations` lists for the keys you need, or set labels on children via other supported paths.

### Unchanged (not part of the propagation tightening)

- **VCT → PVC labeling** (#7910): still the main focus of this PR — additive sync of non-reserved labels onto PVCs.
- **Propagation when `propagate-*` is unset:** no change; only `propagate-labels: *` / `propagate-annotations: *` behavior is tightened.

**Release note (propagation):** With `eck.k8s.alpha.elastic.co/propagate-labels: "*"`, ECK-owned label keys on the parent resource are no longer copied to child resources. This matches existing annotation propagation behavior and avoids duplicating operator-managed identity labels. Custom labels and explicit propagation lists are unchanged. VCT→PVC labeling and disabling propagation entirely are unaffected.

## Summary

- Allow user-defined labels on volume claim templates; propagate them onto existing PVCs during reconciliation.
- Keep reserved ECK label keys protected end-to-end (admission + reconciler strip + sync skip).
- Centralize ECK reserved metadata keys (`metadata/reserved`) for VCT validation, PVC sync, and propagation; tighten `propagate-*: *` so ECK-owned labels are not propagated (annotations aligned).
- Extend validation, webhooks, and unit tests for update paths and edge cases.

## Test plan

<details>
<summary>Manual scenarios executed</summary>

### Scenario 1 — Create-time labels
- PVCs and StatefulSet VCT carry `velero.io/exclude-from-backup` and `test.scenario`.

### Scenario 2 — Add label to existing ES
- Patch accepted; PVC labeled.
- StatefulSet VCT labels preserved empty (immutable).
- Operator logs show no StatefulSet `Forbidden` errors for label-only paths.

### Scenario 3 — Modify existing label
- `team=search` applied on PVC; ES Ready.

### Scenario 4 — Remove labels
- Webhook accepted `labels: {}`.
- PVC retains `team=search` (additive-only).

### Scenario 5 — Label + storage increase
- Label applied; PVC capacity still `1Gi` with `FileSystemResizePending`.

### Scenario 6 — Scale-up after label change
- New PVCs `-1`, `-2` labeled.

### Scenario 7 — Multi-NodeSet
- `tier=hot` / `tier=warm` labels applied to correct PVCs after patch.

### Scenario 9 — Reject storageClass change + label
- Rejected with `PvcImmutableErrMsg`.

### Scenario 10 — Reject claim rename
- Rejected with `PvcImmutableErrMsg`.

### Scenario 11 — Reject storage decrease
- Rejected with storage decrease message.

</details>

**Automated coverage:** table-driven tests in `pvc_expansion_test.go`, `volume/validations/*_test.go`, Elasticsearch `volume_validation_test.go` / `webhook_test.go`, Logstash `validations_test.go` / `webhook_test.go`, plus controller tests touching StatefulSet/VCT behavior where noted in the diff.

### CODE REVIEW

**Comment by prodsecmachine** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scan Engine |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/elastic-cloud-on-k8s/pr-checks/08655031-be66-484f-b407-be59ba5b6704?refs=45fbc4c48caedddbe433266e4865566129771c7e&source=prChecksComment) |
| :white_check_mark: | Licenses| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/elastic-cloud-on-k8s/pr-checks/08655031-be66-484f-b407-be59ba5b6704/license?refs=45fbc4c48caedddbe433266e4865566129771c7e&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

**Comment by naemono** [comment]
I'm going to release this to the team to start reviews, even though BK isn't kicking off the final checks, as GH is having another outage.

**Comment by naemono** [comment]
buildkite test this -f p=gke,t=.*,E2E_TAGS=es -m s=8.19.14

**Comment by naemono** [comment]
buildkite test this -f p=gke,t=.*,E2E_TAGS=es -m s=8.19.14,s=9.3.4,s=7.17.29

**Review by barkbay — COMMENTED** [review]
Label value change on the CR (e.g. `foo=bar` → `foo=bar2`): ECK keeps live VCT on StatefulSet update. CR intent reaches PVCs via `ReconcilePVCsForStatefulSet` → `syncPVCLabels`, then `Update` on each changed PVC (`pvc_expansion.go`).

**1. Will the StatefulSet controller ever reconcile labels?**  
StatefulSet controller cannot patch `volumeClaimTemplates` on an existing StatefulSet. It does not relabel existing PVCs from the template. No conflict with ECK on that path **today**. If Kubernetes allows mutable VCT labels or syncs template labels to PVCs later, we would need to revisit stale VCT preservation and who owns PVC label updates.

**2. Scale-up is eventual consistency.**  
New PVCs come from the **stale** VCT on the StatefulSet. They may miss CR labels until the next ECK reconcile. That is expected and already noted (**scale-up after label change** in the PR), but it kind of reveals that what we do here is a bit "hacky".

File: pkg/controller/common/volume/validations/validations.go [hunk 1]
```diff
@@ -8,19 +8,157 @@ import (
...
+	// eckReservedLabelsRootSubdomain is the shortest DNS subdomain (`k8s.elastic.co`)
+	// reserved for ECK-managed label keys; longer subdomains ending in
+	// `.<eckReservedLabelsRootSubdomain>` (e.g. elasticsearch.k8s.elastic.co,
+	// common.k8s.elastic.co, association.k8s.elastic.co) are reserved too.
+	eckReservedLabelsRootSubdomain = "k8s.elastic.co"
 )
 
+// IsReservedLabelKey reports whether key is owned by ECK and must not be set
+// or modified by users via VolumeClaimTemplate labels. ECK-reserved keys use a
+// label subdomain that equals k8s.elastic.co or ends with .k8s.elastic.co
```
Comment (barkbay) [inline pkg/controller/common/volume/validations/validations.go]: IIUC `IsReservedLabelKey (labels)`  splits on `/` and domain must be exactly `k8s.elastic.co` or end with `.k8s.elastic.co.`

Not sure if it matters but this is not consistent with how we validate annotations in the metadata propagation mechanism: https://github.com/elastic/cloud-on-k8s/blob/32d1a57027f611c1c14c35fd82518925b8633462/pkg/controller/common/annotation/metadata_propagation.go#L23

`eck.k8s.alpha.elastic.co` does not end with .k8s.elastic.co (there is an alpha label in the domain), so not reserved in the VCT.

**Comment by barkbay** [comment]
buildkite test this -f p=gke,t=TestVolume|TestMetadataPropagation|TestAutoscaling -m s=8.19.14,s=9.3.4

**Comment by barkbay** [comment]
Closing for now since this hasn't moved in a while, and a few design points still need discussion (StatefulSet VCT immutability / eventual consistency on scale-up, reserved metadata rules, and the propagate-labels:* behavior change).

Happy for anyone to pick it up and revive it.
